protocols/floodsub/src/layer.rs

```diff
@@ -0,0 +1,277 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use cuckoofilter::CuckooFilter;
+use futures::prelude::*;
+use handler::FloodsubHandler;
+use libp2p_core::nodes::{ConnectedPoint, NetworkBehavior, NetworkBehaviorAction};
+use libp2p_core::{nodes::protocols_handler::ProtocolsHandler, PeerId};
+use protocol::{FloodsubMessage, FloodsubRpc, FloodsubSubscription, FloodsubSubscriptionAction};
+use smallvec::SmallVec;
+use std::{collections::VecDeque, iter, marker::PhantomData};
+use std::collections::hash_map::{DefaultHasher, HashMap};
+use tokio_io::{AsyncRead, AsyncWrite};
+use topic::{Topic, TopicHash};
+
+/// Network behaviour that automatically identifies nodes periodically, and returns information
+/// about them.
+pub struct FloodsubBehaviour<TSubstream> {
+    /// Events that need to be produced outside when polling.
+    events: VecDeque<NetworkBehaviorAction<FloodsubRpc, FloodsubMessage>>,
+
+    /// Peer id of the local node. Used for the source of the messages that we publish.
+    local_peer_id: PeerId,
+
+    /// List of peers the network is connected to, and the topics that they're subscribed to.
+    // TODO: filter out peers that don't support floodsub, so that we avoid hammering them with
+    //       opened substream
+    connected_peers: HashMap<PeerId, SmallVec<[TopicHash; 8]>>,
+
+    // List of topics we're subscribed to. Necessary in order to filter out messages that we
+    // erroneously receive.
+    subscribed_topics: SmallVec<[Topic; 16]>,
+
+    // Sequence number for the messages we send.
+    seq_no: usize,
```

Comment:
The one in `protocols_handler.rs` is for internal purposes only. It's to give an ID to each substream we open.
The `seq_no` here is to give an ID to each floodsub message. It is also broadcasted over the network so that everyone knows which message they have already processed or not.